### PR TITLE
CHAIN-339 tma daily checkins

### PR DESCRIPTION
**Pull request by imykolenko**
Screenshot 2024-06-28 at 3 41 34 PM

### CODE REVIEW

File: backend/src/main/kotlin/co/chainring/core/model/telegram/miniapp/TelegramMiniAppCheckInStreakDailyReward.kt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            ticksUntilLastStrikeGranted == null -> 1 to now
+            // same day
+            (ticksUntilNow - ticksUntilLastStrikeGranted) == 0L -> userEntity.checkInStreakDays to userEntity.lastStreakDayGrantedAt
+            // next day
+            (ticksUntilNow - ticksUntilLastStrikeGranted) == 1L -> userEntity.checkInStreakDays + 1 to now
+            // reset strike
+            else -> 1 to now
+        }
+
+        val config = streakDailyRewards.find { it.day == streakDay } ?: streakDailyRewards.maxBy { it.day }
```
Comment (imykolenko) [inline backend/src/main/kotlin/co/chainring/core/model/telegram/miniapp/TelegramMiniAppCheckInStreakDailyReward.kt:49]: when a 7-day streak is reached, the equivalent of the 7th day's reward will be granted for every subsequent day.

**Review by bholzman — APPROVED** [review]
Looks great, nice job!

File: backend/src/main/kotlin/co/chainring/core/model/telegram/miniapp/TelegramMiniAppUser.kt [hunk 2]
```diff
@@ -99,9 +104,9 @@ class TelegramMiniAppUserEntity(guid: EntityID<TelegramMiniAppUserId>) : GUIDEnt
 
         TelegramMiniAppGameReactionTimeEntity.create(this, reactionTimeMs)
 
-        val totalUsers = TelegramMiniAppGameReactionTimeTable.selectAll().count()
-        val fasterUsers = TelegramMiniAppGameReactionTimeEntity.find { TelegramMiniAppGameReactionTimeTable.reactionTimeMs lessEq reactionTimeMs }.count()
-        val percentile = (((totalUsers - fasterUsers).toDouble() / totalUsers) * 100).toInt()
+        val totalReactions = TelegramMiniAppGameReactionTimeTable.selectAll().count()
+        val fasterReactions = TelegramMiniAppGameReactionTimeEntity.find { TelegramMiniAppGameReactionTimeTable.reactionTimeMs lessEq reactionTimeMs }.count()
```
[the remark this answers is not in the document]
Comment (imykolenko) [reply backend/src/main/kotlin/co/chainring/core/model/telegram/miniapp/TelegramMiniAppUser.kt:108]: Right, that totally makes sense. For efficient percentile calculation a simple "time -> count" table with 50ms time tick (5000 max time / 100) would be a good fit. I'll change approach in a separate PR.

File: backend/src/main/kotlin/co/chainring/apps/api/model/tma/ApiModels.kt [hunk 3]
```diff
@@ -1,19 +1,26 @@
...
     val balance: BigDecimalJson,
     val goals: List<Goal>,
     val gameTickets: Long,
+    val checkInStreak: CheckInStreak,
 ) {
     companion object {
-        fun fromEntity(user: TelegramMiniAppUserEntity): GetUserApiResponse =
-            GetUserApiResponse(
+        fun fromEntity(user: TelegramMiniAppUserEntity): GetUserApiResponse {
+            val dailyReward = TelegramMiniAppCheckInStreak.grantDailyReward(user)
```
Comment (antonyshchenko) [inline backend/src/main/kotlin/co/chainring/apps/api/model/tma/ApiModels.kt:21]: This adds a rather unexpected side-effect to `GetUserApiResponse::fromEntity`. I would suggest moving this somewhere else, for example into `telegramMiniAppSecurity` middleware.

**Pull request merged**